prusti-viper/src/encoder/mir_encoder/mod.rs

```diff
@@ -484,43 +485,93 @@ impl<'p, 'v: 'p, 'tcx: 'v> MirEncoder<'p, 'v, 'tcx> {
-                    "bitwise operations on non-boolean types are not supported"
-                ))
-            }
-            unsupported_op => {
-                return Err(EncodingError::unsupported(format!(
-                    "operation '{:?}' is not supported",
-                    unsupported_op
-                )))
+        match (&left, &right) {
+            (vir::Expr::Const(vir::Const::Float(_), _), _) 
+            | (_, vir::Expr::Const(vir::Const::Float(_), _))            
+                => Ok(match op {
+                    mir::BinOp::Eq => self.encode_const_float_eq_cmp(left, right),
+                    mir::BinOp::Ne => self.encode_const_float_ne_cmp(left, right),
+                    mir::BinOp::Gt => vir::Expr::gt_cmp(left, right),
+                    mir::BinOp::Ge => vir::Expr::ge_cmp(left, right),
+                    mir::BinOp::Lt => vir::Expr::lt_cmp(left, right),
+                    mir::BinOp::Le => vir::Expr::le_cmp(left, right),
+                    mir::BinOp::Add => vir::Expr::add(left, right),
+                    mir::BinOp::Sub => vir::Expr::sub(left, right),
+                    mir::BinOp::Div => vir::Expr::div(left, right),
+                    mir::BinOp::Mul => vir::Expr::mul(left, right),
+                    unsupported_op => {
+                        return Err(EncodingError::unsupported(format!(
+                            "operation '{:?}' is not supported for floats",
+                            unsupported_op
+                        )))
+                    }
+                }),
+            
+            _ => {
+                let is_bool = ty.kind() == &ty::TyKind::Bool;
+                Ok(match op {
+                    mir::BinOp::Eq => vir::Expr::eq_cmp(left, right),
+                    mir::BinOp::Ne => vir::Expr::ne_cmp(left, right),
+                    mir::BinOp::Gt => vir::Expr::gt_cmp(left, right),
+                    mir::BinOp::Ge => vir::Expr::ge_cmp(left, right),
+                    mir::BinOp::Lt => vir::Expr::lt_cmp(left, right),
+                    mir::BinOp::Le => vir::Expr::le_cmp(left, right),
+                    mir::BinOp::Add => vir::Expr::add(left, right),
+                    mir::BinOp::Sub => vir::Expr::sub(left, right),
+                    mir::BinOp::Rem => vir::Expr::rem(left, right),
+                    mir::BinOp::Div => vir::Expr::div(left, right),
+                    mir::BinOp::Mul => vir::Expr::mul(left, right),
+                    mir::BinOp::BitAnd if is_bool => vir::Expr::and(left, right),
+                    mir::BinOp::BitOr if is_bool => vir::Expr::or(left, right),
+                    mir::BinOp::BitXor if is_bool => vir::Expr::xor(left, right),
+                    mir::BinOp::BitAnd => vir::Expr::bit_and(left, right),
+                    mir::BinOp::BitOr => vir::Expr::bit_or(left, right),
+                    mir::BinOp::BitXor => vir::Expr::bit_xor(left, right),
+                    mir::BinOp::Shl => unimplemented!(),
```

Comment:
Please remove all `unimplemented!`. If something is not implemented, it should report a meaningful `Err(EncodingError::unsupported(...))`.